include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,3312 @@
+    crab::ScopedCrabStats __st__(domain_name() + ".widening");
+
+    // Trivial cases first: we don't cover cases where one operand is
+    // top because is_top() calls the base domain which we don't know
+    // whether it will perform some normalization or not.
+    if (is_bottom()) {
+      return o;
+    } else if (o.is_bottom()) {
+      return *this;
+    }
+
+    CRAB_LOG("object", crab::outs()
+                           << "Widening " << *this << " and " << o << " =\n");
+
+    object_domain_t res(
+        std::move(join_or_widening(*this, o, false /*is widen*/)));
+
+    CRAB_LOG("object", crab::outs() << res << "\n");
+    return res;
+  }
+
+  object_domain_t
+  widening_thresholds(const object_domain_t &o,
+                      const thresholds<number_t> &thresholds) const override {
+    crab::CrabStats::count(domain_name() + ".count.widening");
+    crab::ScopedCrabStats __st__(domain_name() + ".widening");
+
+    // Trivial cases first: we don't cover cases where one operand is
+    // top because is_top() calls the base domain which we don't know
+    // whether it will perform some normalization or not.
+    if (is_bottom()) {
+      return o;
+    } else if (o.is_bottom()) {
+      return *this;
+    }
+
+    CRAB_LOG("object", crab::outs() << "Widening with threshold " << *this
+                                    << " and " << o << " =\n");
+
+    object_domain_t res(
+        std::move(join_or_widening(*this, o, false /*is widen*/)));
+
+    CRAB_LOG("object", crab::outs() << res << "\n");
+    return res;
+  }
+
+  object_domain_t operator&&(const object_domain_t &o) const override {
+    crab::CrabStats::count(domain_name() + ".count.narrowing");
+    crab::ScopedCrabStats __st__(domain_name() + ".narrowing");
+
+    if (is_bottom() || o.is_top()) {
+      return *this;
+    } else if (o.is_bottom() || is_top()) {
+      return o;
+    }
+
+    CRAB_LOG("object", crab::outs()
+                           << "Narrowing " << *this << " and " << o << " =\n");
+
+    object_domain_t res(
+        std::move(meet_or_narrowing(*this, o, false /*is narrow*/)));
+
+    CRAB_LOG("object", crab::outs() << res << "\n");
+    return res;
+  }
+
+  /***************** Regions and reference operations *****************/
+
+  // Initialize a region
+  void region_init(const variable_t &rgn) override {
+    crab::CrabStats::count(domain_name() + ".count.region_init");
+    crab::ScopedCrabStats __st__(domain_name() + ".region_init");
+
+    ERROR_IF_NOT_REGION(rgn, __LINE__);
+
+    if (is_bottom()) {
+      return;
+    }
+
+    // for now skip analysis for unknown region
+    if (is_unknown_region(rgn)) {
+      return;
+    }
+
+    // REDUCTION: perform reduction
+    perform_reduction();
+
+    if (get_obj_id(rgn) == boost::none) {
+      // if a region does not belong to an object, treat it as an object
+      // treat region as a field, as well as an object id
+      update_fields_id_map(rgn, rgn);
```

Comment:
In these 2 functions you are doing `(*m_flds_id_map).find(rgn)` twice. See if this can be improved